tpm/src/socket.rs

```diff
@@ -0,0 +1,160 @@
+    #[error("Cannot connect to tpm Socket even after retrying for 10 secs")]
+    ConnectToSocket(#[source] anyhow::Error),
+    #[error("Failed to read from socket")]
+    ReadFromSocket(#[source] anyhow::Error),
+    #[error("Failed to write to socket")]
+    WriteToSocket(#[source] anyhow::Error),
+}
+type Result<T> = anyhow::Result<T, Error>;
+
+#[derive(PartialEq)]
+enum SocketDevState {
+    Disconnected,
+    Connecting,
+    Connected,
+}
+
+pub struct SocketDev {
+    state: SocketDevState,
+    stream: Option<UnixStream>,
+    // Fd sent to swtpm process for Data Channel
+    write_msgfd: RawFd,
+    // Data Channel used by Cloud-Hypervisor
+    data_fd: RawFd,
+    // Control Channel used by Cloud-Hypervisor
+    ctrl_fd: RawFd,
+}
+
+impl Default for SocketDev {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl SocketDev {
+    pub fn new() -> Self {
+        Self {
+            state: SocketDevState::Disconnected,
+            stream: None,
+            write_msgfd: -1,
+            ctrl_fd: -1,
+            data_fd: -1,
+        }
+    }
+
+    pub fn init(&mut self, path: String) -> Result<()> {
+        self.connect(&path)?;
+        Ok(())
+    }
+
+    pub fn connect(&mut self, socket_path: &str) -> Result<()> {
+        self.state = SocketDevState::Connecting;
+
+        let now = Instant::now();
+
+        // Retry connecting for 10 secs
+        loop {
+            match UnixStream::connect(socket_path) {
+                Ok(s) => {
+                    let fd = s.as_raw_fd();
+                    self.ctrl_fd = fd;
+                    self.stream = Some(s);
+                    self.state = SocketDevState::Connected;
+                    debug!("Connected to tpm socket path : {:?}", socket_path);
+                    return Ok(());
+                }
+                Err(_e) => {}
+            };
+            sleep(Duration::from_millis(1000));
+
+            if now.elapsed().as_secs() >= 10 {
+                break;
+            }
+        }
+        Err(Error::ConnectToSocket(anyhow!(
+            "Failed to connect to tpm Socket"
+        )))
+    }
+
+    pub fn set_datafd(&mut self, fd: RawFd) {
+        self.data_fd = fd;
+    }
+
+    pub fn set_msgfd(&mut self, fd: RawFd) {
+        self.write_msgfd = fd;
+    }
+
+    pub fn send_full(&self, buf: &[u8]) -> Result<usize> {
+        let write_fd = self.write_msgfd;
+
+        let size = self
+            .stream
+            .as_ref()
+            .unwrap()
+            .send_with_fd(buf, write_fd)
+            .map_err(|e| {
+                Error::WriteToSocket(anyhow!("Failed to write to Socket. Error: {:?}", e))
+            })?;
+
+        Ok(size)
+    }
+
+    pub fn write(&mut self, buf: &[u8]) -> Result<usize> {
+        if self.stream.is_none() {
+            return Err(Error::WriteToSocket(anyhow!(
+                "Stream for tpm socket was not initialized"
+            )));
+        }
+        let res = match self.state {
```

Comment:
You could simplify this with `if matches!(self.state,SocketDevState::Connected)`

Reply:
updated!